src/coprocessor/codec/mysql/duration.rs

```diff
@@ -460,35 +671,52 @@ impl Display for Duration {
             self.minutes(),
             self.secs()
         )?;
-        if self.fsp > 0 {
-            write!(formatter, ".")?;
-            let nanos = self.dur.subsec_nanos() / (10u32.pow(NANO_WIDTH - u32::from(self.fsp)));
-            write!(formatter, "{:01$}", nanos, self.fsp as usize)?;
+
+        let fsp = usize::from(self.fsp());
+
+        if self.fsp() > 0 {
+            write!(
+                formatter,
+                ".{:0width$}",
+                self.micros() / TEN_POW[6 - fsp],
+                width = fsp
+            )?;
         }
+
         Ok(())
     }
 }
 
 impl PartialEq for Duration {
     fn eq(&self, dur: &Duration) -> bool {
-        self.neg == dur.neg && self.dur.eq(&dur.dur)
+        let (mut a, mut b) = (*self, *dur);
+        a.set_fsp(0);
+        b.set_fsp(0);
+        a.0 == b.0
     }
 }
 
 impl std::hash::Hash for Duration {
     fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
-        self.neg.hash(state);
-        self.dur.hash(state);
+        let mut inner = *self;
+        inner.set_fsp(0);
+        inner.to_bits().hash(state)
     }
 }
 
 impl PartialOrd for Duration {
     fn partial_cmp(&self, dur: &Duration) -> Option<Ordering> {
-        Some(match (self.neg, dur.neg) {
-            (true, true) => dur.dur.cmp(&self.dur),
+        let mut a = *self;
+        let mut b = *dur;
+
+        a.set_fsp(0);
+        b.set_fsp(0);
+
+        Some(match (a.get_neg(), b.get_neg()) {
```

Comment:
I believe we can simply bit of neg, then compare a.0 and b.0 directly. (or better way is to store 1 as positive and store 0 as negative).